File: sections/69_master_structures_trees_mixed/problems/f_subtree_frequency_profile/solution.cpp
#include <bits/stdc++.h>
using namespace std;
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n;
  if (!(cin >> n)) return 0;
  vector<int> color(n);
  for (int& x : color) cin >> x;
  vector<vector<int>> g(n);
  for (int i = 1, u, v; i < n; i++) {
    cin >> u >> v;
    --u;
    --v;
    g[u].push_back(v);
    g[v].push_back(u);
  }
  vector<int> parent(n, -1), order{0};
  for (int i = 0; i < (int)order.size(); i++) {
    int u = order[i];
    for (int v : g[u])
      if (v != parent[u]) parent[v] = u, order.push_back(v);
  }
  vector<unordered_map<int, int>*> maps(n);
  vector<int> best(n), ways(n);
  for (int at = n - 1; at >= 0; at--) {
    int u = order[at], base = -1;
    for (int v : g[u])
      if (parent[v] == u && (base < 0 || maps[v]->size() > maps[base]->size()))
        base = v;
    if (base < 0)
      maps[u] = new unordered_map<int, int>();
    else
      maps[u] = maps[base], best[u] = best[base], ways[u] = ways[base];
    auto add = [&](int c, int amount) {
      int value = ((*maps[u])[c] += amount);
      if (value > best[u])
        best[u] = value, ways[u] = 1;
      else if (value == best[u])
        ways[u]++;
    };
    for (int v : g[u])
      if (parent[v] == u && v != base) {
        for (auto [c, x] : *maps[v]) add(c, x);
        delete maps[v];
      }
    add(color[u], 1);
  }
  for (int u = 0; u < n; u++) cout << best[u] << ' ' << ways[u] << '\n';
  delete maps[0];
}

File: sections/69_master_structures_trees_mixed/problems/f_subtree_frequency_profile/solution.py
import sys


def main():
    d = list(map(int, sys.stdin.buffer.read().split()))
    if not d:
        return
    n = d[0]
    colors = d[1 : n + 1]
    z = n + 1
    g = [[] for _ in range(n)]
    for _ in range(n - 1):
        u = d[z] - 1
        v = d[z + 1] - 1
        z += 2
        g[u].append(v)
        g[v].append(u)
    parent = [-1] * n
    order = [0]
    for u in order:
        for v in g[u]:
            if v != parent[u]:
                parent[v] = u
                order.append(v)
    maps = [None] * n
    best = [0] * n
    ways = [0] * n
    for u in reversed(order):
        children = [v for v in g[u] if parent[v] == u]
        base = max(children, key=lambda v: len(maps[v]), default=-1)
        if base < 0:
            current = {}
            maximum = count = 0
        else:
            current = maps[base]
            maximum = best[base]
            count = ways[base]

        def add(color, amount):
            nonlocal maximum, count
            value = current.get(color, 0) + amount
            current[color] = value
            if value > maximum:
                maximum = value
                count = 1
            elif value == maximum:
                count += 1

        for v in children:
            if v != base:
                for color, amount in maps[v].items():
                    add(color, amount)
                maps[v].clear()
        add(colors[u], 1)
        maps[u] = current
        best[u] = maximum
        ways[u] = count
    print("\n".join(f"{best[u]} {ways[u]}" for u in range(n)))


if __name__ == "__main__":
    main()
